src/libraries/Common/src/System/HexConverter.cs

```diff
@@ -84,11 +90,75 @@ public static void ToCharsBuffer(byte value, Span<char> buffer, int startingInde
             buffer[startingIndex] = (char)(packedResult >> 8);
         }
 
+#if SYSTEM_PRIVATE_CORELIB
+    private static void EncodeToUtf16_Ssse3(ReadOnlySpan<byte> bytes, Span<char> chars, Casing casing)
+    {
+        Debug.Assert(bytes.Length >= 4);
+        nint pos = 0;
+
+        Vector128<byte> shuffleMask = Vector128.Create(
+            0xFF, 0xFF, 0, 0xFF, 0xFF, 0xFF, 1, 0xFF,
+            0xFF, 0xFF, 2, 0xFF, 0xFF, 0xFF, 3, 0xFF);
+
+        Vector128<byte> asciiTable = (casing == Casing.Upper) ?
+            Vector128.Create((byte)'0', (byte)'1', (byte)'2', (byte)'3',
+                             (byte)'4', (byte)'5', (byte)'6', (byte)'7',
+                             (byte)'8', (byte)'9', (byte)'A', (byte)'B',
+                             (byte)'C', (byte)'D', (byte)'E', (byte)'F') :
+            Vector128.Create((byte)'0', (byte)'1', (byte)'2', (byte)'3',
+                             (byte)'4', (byte)'5', (byte)'6', (byte)'7',
+                             (byte)'8', (byte)'9', (byte)'a', (byte)'b',
+                             (byte)'c', (byte)'d', (byte)'e', (byte)'f');
+
+        do
+        {
+            // Read 32bits from "bytes" span at "pos" offset
+            uint block = Unsafe.ReadUnaligned<uint>(
+                ref Unsafe.Add(ref MemoryMarshal.GetReference(bytes), pos));
+
+            // Calculate nibbles
+            Vector128<byte> lowNibbles = Ssse3.Shuffle(
+                Vector128.CreateScalarUnsafe(block).AsByte(), shuffleMask);
+            Vector128<byte> highNibbles = Sse2.ShiftRightLogical(
+                Sse2.ShiftRightLogical128BitLane(lowNibbles, 2).AsInt32(), 4).AsByte();
+
+            // Lookup the hex values at the positions of the indices
+            Vector128<byte> indices = Sse2.And(
+                Sse2.Or(lowNibbles, highNibbles), Vector128.Create((byte)0xF));
+            Vector128<byte> hex = Ssse3.Shuffle(asciiTable, indices);
+
+            // The high bytes (0x00) of the chars have also been converted
+            // to ascii hex '0', so clear them out.
+            hex = Sse2.And(hex, Vector128.Create((ushort)0xFF).AsByte());
```

Comment:
Idea for micro-optimization.

`Vector.Create` creates the constant, that needs to be read from memory. With the other "constants" used here, it's likely they will be brought together into L1D, so the cost should be low. 

But it still evicts a cacheline from L1D's set, so one could avoid this load by using bit-shifting for the masking. Something like
```c#
// (pseudo code style)
tmp = Sse2.ShiftLeftLogical(vec, 8)
Sse2.ShiftRightLogical(tmp, 8)
```
Latency and throughput for the shift is good, but this introduces a register dependency.

In micro-benchmarks the L1D will be hot, so it won't harm here, but in real usages that's likely not the case, so avoiding the load may be a plus.

TBH I don't know if it's worth it and so far it's more theory...

Reply:
Nice idea! However, I don't see any differences in benchmarks (even slightly slower).
```csharp
Sse2.And(hex, Vector128.Create((ushort)0xFF))
```
is emitted as `vpand    xmm0, xmm0, xmmword ptr [reloc @RWD00]` (memory load without an additional register)
It probably makes sense to hoist them from loop too but it also doesn't affect the benchmarks (almost).

Reply:
> doesn't affect the benchmarks

L1D is hot, so I expected this. 
For real world usage I doubt the benchmarks will show a measurable difference too.
But according the theory a different cacheline could be kept in the sets of L1D, as it's not evicted, which could be good in the overall.

To show this in a benchmark one would need the address of the `@RWD00` so the cache set is known (with cpu data from cpu-z, etc.). Then load from the address and other data(s), that "fight" for the same cache set, so this set is evicted constantly. But this is almost impossible to do reliable in a benchmark.